### PR TITLE
Small seam to expose the Contact CRUD through existing user interface

### PR DESCRIPTION
**Pull request by zspencer**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,6 +87,15 @@ Before you can interact with the Nylas API, you need to register for the Nylas D
 Generally, you should store your App ID and Secret into environment variables to avoid adding them to source control. That said, in the example project and code snippets below, the values were added to `config/environments/development.rb` for convenience.
 
 
+### V2
+
+The V2 Nylas API includes support for CRUD operations on Contacts, expands contacts to include more data, and improves contact filtering. It is backwards incompatible with the V1 Contacts API.
+
+The following examples are from [examples/v2.rb](examples/v2.rb)
+
+{include:file:examples/v2.rb}
```
Comment (zspencer) [inline README.md]: When you look at the docs on rubydoc.info, the file `examples/v2.rb` will be inline here. For those viewing on Github, they will be linked to the file above.

**Comment by zspencer** [comment]
@khamidou - 

When I send the following JSON to the contacts contacts API, it doesn't persist the street address or email addresses:

```
{"given_name"=>"Dulce", "surname"=>"Littel", "email_addresses"=>[{"type"=>"personal", "email"=>"dena.jacobson@example.org"}], "physical_addresses"=>[{"format"=>"structured", "street_address"=>"49810 Shanahan Terrace", "postal_code"=>"65696-4425", "state"=>"Wyoming", "city"=>"New Bradley", "country"=>"America"}], "phone_numbers"=>[{"type"=>"mobile", "number"=>"1-672-224-1553"}]}
```

When I retrieve the data later, tis is what I get:

```
{:account_id=>"30zipv27dtrsnkleg59mprw5p", :birthday=>{:date=>nil, :object=>"date"}, :company_name=>nil, :email_addresses=>[], :given_name=>"Dulce", :id=>"f60uxbo351u1crqiw38quuh4", :im_addresses=>[], :job_title=>nil, :manager_name=>nil, :middle_name=>nil, :nickname=>nil, :notes=>nil, :object=>"contact", :office_location=>nil, :phone_numbers=>[{:number=>"1-672-224-1553", :type=>nil}], :physical_addresses=>[{:country=>"America", :format=>"structured", :postal_code=>"65696-4425", :state=>"Wyoming", :street_address=>nil, :type=>nil}], :suffix=>nil, :surname=>"Littel", :web_page=>nil, :web_pages=>[]}
```

**Comment by zspencer** [comment]
Further, for every 20 random contacts I create, 2 of them fail due to internal errors:

```

E, [2017-12-03T17:47:31.710879 #19916] ERROR -- : {:physical_addresses=>[{:format=>"structured", :type=>"work", :street_address=>"59043 Aryanna Rest", :postal_code=>"69116", :city=>"Sageton", :state=>"California", :country=>"America"}], :phone_numbers=>[{:type=>nil, :number=>"(623) 087-5542"}], :email_addresses=>[{:type=>"personal", :email=>"wilford@example.com"}], :given_name=>"Stephen", :surname=>"Sanford"}
E, [2017-12-03T17:47:31.711130 #19916] ERROR -- : An internal error occured. If this issue persists, please contact support@nylas.com and include this request_uid: 6BD29C22:EE12_0A4D0050:01BB_5A24A933_1E8F357:2C61
E, [2017-12-03T17:47:33.533148 #19916] ERROR -- : {:physical_addresses=>[{:format=>"structured", :type=>"work", :street_address=>"898 Jacobs Crest", :postal_code=>"94604-5721", :city=>"Bechtelarshire", :state=>"Minnesota", :country=>"America"}], :phone_numbers=>[{:type=>nil, :number=>"485.545.3443"}], :email_addresses=>[{:type=>nil, :email=>"bertrand@example.org"}], :given_name=>"Georgette", :surname=>"Waelchi"}
E, [2017-12-03T17:47:33.533291 #19916] ERROR -- : An internal error occured. If this issue persists, please contact support@nylas.com and include this request_uid: 6BD29C22:EE17_0A4D0AED:01BB_5A24A934_1E05B6F:3C77
```

File: lib/nylas/logging.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    def self.logger
+      return @logger if @logger
+      @logger = Logger.new(STDOUT)
+      @logger.level = level
+      @logger
+    end
+
+    def self.level
+      Logger.const_get(ENV['NYLAS_LOG_LEVEL'] || :WARN)
```
Comment (khamidou) [inline lib/nylas/logging.rb]: 👍

File: lib/nylas/registry.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+module Nylas
+  class Registry
```
Comment (khamidou) [inline lib/nylas/registry.rb]: What is this registry for?

File: lib/nylas/v2/constraints.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+module Nylas
+  module V2
+    class Constraints
+      attr_accessor :where, :limit, :offset, :view, :per_page
```
Comment (khamidou) [inline lib/nylas/v2/constraints.rb]: Is this the only way to define filters? How are people going to define custom filters, say for example for filtering contacts on a specific address?

File: lib/nylas/v2/model/attribute_definition.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+module Nylas
+  module V2
+    module Model
+      class AttributeDefinition
+        extend Forwardable
```
Comment (khamidou) [inline lib/nylas/v2/model/attribute_definition.rb]: What does `Forwardable` do?

File: lib/nylas/registry.rb (hunk 3 above)
Comment (zspencer) [reply lib/nylas/registry.rb]: Sure, this is a small implementation of the Registry pattern, we're using the Registry class in a couple places:

1. To register the different custom types that data gets serialized in and out of.
2. To register the attributes a given model has

We're using it to make sure that a reasonable error message is provided if we attempt to set or retrieve data that doesn't actually map accurately; as Ruby hashes don't have mechanisms for specifying that directly, and I didn't want to re-open classes at runtime.

File: lib/nylas/v2/constraints.rb (hunk 4 above)
Comment (zspencer) [reply lib/nylas/v2/constraints.rb]: This is so you can have syntax like: `api.contacts.where(street_address: "123 N West St").limit(10).offset(25)` to get the next 10 people who are in `123 N West St`

This is part of the providing an ActiveModel/ActiveRecord/ActiveResource type syntax for users of the 4.0 SDK.

File: lib/nylas/v2/model/attribute_definition.rb (hunk 5 above)
Comment (zspencer) [reply lib/nylas/v2/model/attribute_definition.rb]: Forwardable is a Ruby module for delegating methods from Instance A to Attribute B. It's a useful tool when relying on composition for organization of responsibility.

**Comment by zspencer** [comment]
@khamidou - yea this is a big diff. I took on three challenges at once which wasn't really a great idea, but seemed like the best way to move forward quickly with the V2 functionality while driving the SDK interface forward.

The most important file to review is `examples/v2.rb` which demonstrates the syntax of the user interface for querying the contacts collection and for updating/saving/deleting contacts.

We went back and forth a bit on how we could go about making sure this is an incremental upgrade without requiring clients to reimplement things, and the path we settled on when you Tomasz and I were talking was:

1. Leave as much of the existing interface in place as possible, so that people on V1 of the API don't/barely have to change anything
2. Get the V2 Contact API working with the ActiveModel/ActiveRecord/ActiveResource compatible interface, as people will have to update their code to work with the new Contact schema anyway.
3.  Start filling out the application with models that conform to the ActiveModel/ActiveResource/ActiveRecord compliant interface without requiring people to upgrade, mark the non-conforming models as deprecated so new implementors are discouraged from using them.
4. When we release nylas-ruby 5.0 (in a while, so that clients have time to migrate), we'll delete the non-conforming models.

We want to keep the V2 namespace around for when you sprout a new endpoint that is only available to V2 customers. The V2 namespace will likely only include the Model definitions for V2-specific endpoints, I was hesitant to pull them up to the top level just yet because I didn't want to shovel around the code that will be deprecated until I've got the new stuff all working. This was a big enough change without juggling that (IMO).

File: examples/v2.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+
+demonstrate { api.contacts.count }
+
+# Searching!
+# demonstrate { api.contacts.where(email: loaded_contacts[8].email_addresses.first.email).limit(3).map(&:to_h) }
+demonstrate { api.contacts.where(state: loaded_contacts[3].physical_addresses.first&.state).limit(2).map(&:to_h) }
+demonstrate { api.contacts.where(country: loaded_contacts[5].physical_addresses.first&.country).limit(2).map(&:to_h) }
+demonstrate { api.contacts.where(phone_number: loaded_contacts[12].phone_numbers.first&.number).limit(2).map(&:to_h) }
+demonstrate { api.contacts.where(street_address: loaded_contacts[13].physical_addresses.first&.street_address).limit(2).map(&:to_h) }
```
Comment (zspencer) [inline examples/v2.rb]: Make sure this list has examples for each filter listed in the docs.

**Pull request merged**